File: java/src/main/java/it/giannibombelli/wsc2026/booking/infrastructure/InMemoryBookingEventBus.java
package it.giannibombelli.wsc2026.booking.infrastructure;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.booking.domain.events.BookingEvent;
import it.giannibombelli.wsc2026.common.application.events.EventBus;
import it.giannibombelli.wsc2026.common.application.events.EventSubscriber;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executor;

public final class InMemoryBookingEventBus implements EventBus<BookingEvent> {

    private final Map<Class<? extends BookingEvent>, List<EventSubscriber<? extends BookingEvent>>> subscribers = new ConcurrentHashMap<>();
    private final Executor executor;

    public InMemoryBookingEventBus(Executor executor) {
        Require.requireDependency(executor, "executor");
        this.executor = executor;
    }

    @Override
    public void publish(BookingEvent event) {
        Require.requireArgument(event, "event");

        List<EventSubscriber<? extends BookingEvent>> matching = subscribers.entrySet().stream()
            .filter(entry -> entry.getKey().isAssignableFrom(event.getClass()))
            .flatMap(entry -> entry.getValue().stream())
            .toList();

        executor.execute(() -> matching.forEach(subscriber -> notify(subscriber, event)));
    }

    @SuppressWarnings("unchecked")
    private <E extends BookingEvent> void notify(EventSubscriber<E> subscriber, BookingEvent event) {
        subscriber.on((E) event);
    }

    @Override
    public void subscribe(Class<? extends BookingEvent> eventType, EventSubscriber<? extends BookingEvent> subscriber) {
        Require.requireArgument(eventType, "eventType");
        Require.requireArgument(subscriber, "subscriber");
        subscribers.computeIfAbsent(eventType, k -> new CopyOnWriteArrayList<>()).add(subscriber);
    }
}

File: typescript/src/booking/infrastructure/inMemoryBookingEventBus.ts
// Event bus in-memory specializzato per BookingEvent.
// Supporta dispatch sincrono nei test e configurabile in produzione tramite Executor.

import type { EventBus } from '@/common/application/events/eventBus.js';
import type { EventSubscriber } from '@/common/application/events/eventSubscriber.js';
import type { BookingEvent } from '../domain/events/bookingEvent.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export type Executor = (task: () => void) => void;

export class InMemoryBookingEventBus implements EventBus<BookingEvent> {
  private readonly _subscribers = new Map<string, EventSubscriber<BookingEvent>[]>();
  private readonly _executor: Executor;

  constructor(executor: Executor) {
    requireDependency(executor, "executor");
    this._executor = executor;
  }

  publish(event: BookingEvent): void {
    requireArgument(event, 'event');

    const list = this._subscribers.get(event.kind) ?? [];
    this._executor(() => list.forEach((subscriber) => subscriber.on(event)));
  }

  subscribe(eventKind: string, subscriber: EventSubscriber<BookingEvent>): void {
    requireArgument(eventKind, 'eventKind');
    requireArgument(subscriber, 'subscriber');
    const list = this._subscribers.get(eventKind) ?? [];
    list.push(subscriber);
    this._subscribers.set(eventKind, list);
  }
}
